onnxruntime/core/common/threadpool.cc

```diff
@@ -402,29 +402,45 @@ void ThreadPool::ParallelForFixedBlockSizeScheduling(const std::ptrdiff_t total,
     return;
   }
 
-  // Split the work across threads in the pool.  Each work item will run a loop claiming iterations,
-  // hence we need at most one for each thread, even if the numberof blocks of iterations is larger.
   auto d_of_p = DegreeOfParallelism(this);
-  auto num_blocks = total / block_size;
-  auto num_threads_inc_main = NumThreads() + 1;
-  int num_work_items = static_cast<int>(std::min(static_cast<std::ptrdiff_t>(num_threads_inc_main), num_blocks));
-  assert(num_work_items > 0);
-
-  LoopCounter lc(total, d_of_p, block_size);
-  std::function<void(unsigned)> run_work = [&](unsigned idx) {
-    unsigned my_home_shard = lc.GetHomeShard(idx);
-    unsigned my_shard = my_home_shard;
-    uint64_t my_iter_start, my_iter_end;
-    while (lc.ClaimIterations(my_home_shard, my_shard, my_iter_start, my_iter_end)) {
-      fn(static_cast<std::ptrdiff_t>(my_iter_start),
-         static_cast<std::ptrdiff_t>(my_iter_end));
-    }
-  };
-
-  // Run the work in the thread pool (and in the current thread).  Synchronization with helping
-  // threads is handled within RunInParallel, hence we can deallocate lc and other state captured by
-  // run_work.
-  RunInParallel(run_work, num_work_items, block_size);
+  if (thread_options_.dynamic_block_base_ <= 0) {
+    // Split the work across threads in the pool.  Each work item will run a loop claiming iterations,
+    // hence we need at most one for each thread, even if the number of blocks of iterations is larger.
+    auto num_blocks = total / block_size;
+    auto num_threads_inc_main = NumThreads() + 1;
+    int num_work_items = static_cast<int>(std::min(static_cast<std::ptrdiff_t>(num_threads_inc_main), num_blocks));
+    assert(num_work_items > 0);
+
+    LoopCounter lc(total, d_of_p, block_size);
+    std::function<void(unsigned)> run_work = [&](unsigned idx) {
+      unsigned my_home_shard = lc.GetHomeShard(idx);
+      unsigned my_shard = my_home_shard;
+      uint64_t my_iter_start, my_iter_end;
+      while (lc.ClaimIterations(my_home_shard, my_shard, my_iter_start, my_iter_end)) {
+        fn(static_cast<std::ptrdiff_t>(my_iter_start),
+           static_cast<std::ptrdiff_t>(my_iter_end));
+      }
+    };
+    // Run the work in the thread pool (and in the current thread).  Synchronization with helping
+    // threads is handled within RunInParallel, hence we can deallocate lc and other state captured by
+    // run_work.
+    RunInParallel(run_work, num_work_items, block_size);
+  } else {
+    int num_of_blocks = d_of_p * thread_options_.dynamic_block_base_;
+    std::ptrdiff_t block_size_base{1};
+    while (block_size_base * num_of_blocks < total) block_size_base <<= 1;
+    std::atomic<ptrdiff_t> iter{0};
+    std::function<void(unsigned)> run_work = [&](unsigned) {
+      std::ptrdiff_t start{0};
+      std::ptrdiff_t b = block_size_base;
+      while ((start = iter.fetch_add(b, std::memory_order_relaxed)) < total) {
+        fn(start, std::min(start + b, total));
```

Comment:
One consequence of using fetch_add on a single counter here is that we _lose_ the ability to cause the same thread to tend to be allocated the same work across a series of loops (e.g. thread 0 runs iterations 0..9, thread 1 runs iterations 10..19, etc).  That was important for cache locality in some models -- I think it was one using GRU where we first saw that, and e-mail discussion w Tracy and maybe Du Li?

Could we pushing the batch-size selection _inside_ LoopCounter, so that we can continue using the same sharding work there to provide this locality?

That will also help avoid needing two different kinds of loop -- we would just have a change _inside_ LoopCounter to manage the dynamic batch size selection.

Reply:
With latest commits I am reusing the LoopCounter for better cache locality.